index.bs

```diff
@@ -2015,14 +1988,107 @@ Issue: This needs to be generalised to work with realms too.
+      1. If |top-level traversable context ids| is an empty [=/set=]:
 
-1. [=Try=] to [=update the event map=] with |session|,
-   |list of event names|, |list of contexts| and enabled false.
+        1. |=list/Remove=| all items [=list/contains|contained=] in |event names| from |subscription|'s [=subscription/event names=].
+
+        1. If |subscription|'s [=subscription/event names=] is not empty:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+      1. Otherwise:
+
+        1. If |subscription| is [=subscription/global=]:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+        1. Otherwise:
+
+          Note: unsubscribe by contexts is deprecated and will be removed in the future versions.
+
+          1. Let |event map| be an empty [=/map=].
+
+          1. For each |event name| in |subscription|'s [=subscription/event names=]:
+
+              1. Set |event map|[|event name|] to [=set/clone=] of |subscription|'s [=subscription/top-level traversable ids=].
+
+          1. For each |event name| in |event names|:
+
+              1. If |event map|[|event name|] does not exist,
+                 continue.
+
+              1. Set |event map|[|event name|] to the [=set/difference=]
+                 between |event map|[|event name|] and |top-level traversable context ids|.
+
+              1. If |event map|[|event name|] is an empty [=/set=],
+
+                1. [=map/Remove=] |event map|[|event name|].
+
+          1. For each |event name| → |remaining top-level traversable ids| in |event map|:
+
+            1. Let |partial subscription| be a [=subscription=] with
+               [=subscription/subscription id=] to |subscription|'s [=subscription/subscription id=],
+               [=subscription/event names=] set to a new [=/set=] containing |event name|,
+               [=subscription/top-level traversable ids=] set to |remaining top-level traversable ids|.
+
+            1. [=list/append=] |partial subscription| to |new subscriptions|.
+
+   1. Set |session|'s [=subscriptions=] to |new subscriptions|.
+
+1. Otherwise:
+
+   1. Let |subscriptions| be a [=/set=] of |command parameters|[<code>subscriptions</code>].
+
+   1. Let |subscriptions to remove| be an empty [=/set=].
+
+   1. Let |found subscription ids| be an empty [=/set=].
+
+   1. For each |subscription| in |session|'s [=subscriptions=]:
+
+      1. If |subscriptions| [=set/contains=] |subscription|'s [=subscription/subscription id=]:
+
+         1. [=set/Append=] |subscription| to |subscriptions to remove|.
+
+         1. [=set/Append=] |subscription|'s [=subscription/subscription id=] to |found subscription ids|.
+
+   1. Let |not found subscriptions| be [=set/difference=] between |subscriptions| and |found subscription ids|.
+
+   1. Set |not found subscriptions| be [=set/difference=] between |found subscription ids| and |session|'s [=internally removed subscription ids=].
```

Comment:
The difference between |found subscription ids| and |session|'s [=internally removed subscription ids=] contains all the internally removed subscriptions. 

We need to iterate through the |not found subscriptions| and remove each [=internally removed subscription ids=] from there.

Reply:
UPD: my bad, the "difference" is not commutative.